### ISSUE TITLE
Submit jobs to queue locally

### ISSUE THREAD

**Issue by jean997** [stephenslab/dsc#131]
There is currently no way to submit jobs to a queue locally although it is possible to submit jobs to a queue on a remote host. This functionality is important because a) for long running jobs it may be easier to have the parent job running on a remote server and b) it is often desirable to store the data remotely as well.

**Comment by pcarbo** [comment]
@jean997 Can you give a use case for each? Why not have DSC run on the remote host?

**Comment by jean997** [comment]
Sure. I'd like to run simulations from rcc because a) I want to be able to turn my laptop off and take it with me and b) I would like all my simulations and code to be stored on rcc with the rest of my project data. I realized that results currently would get stored on the remote host but it is just organizationally easier to have everything in the same place. It seems unnecessary to create a new directory with copies of code that needs to be synced on a separate computer just so that I can use a cluster.

**Comment by jean997** [comment]
Since it would use the version of DSC I have installed on rcc it would also mean that I could access my project from any computer, even one that doesn't have DSC installed. That could be very convenient as well.

**Comment by gaow** [comment]
Sorry @jean997 and @pcarbo i did not noticed a ticket has been submitted when I posted mine #132 . I'll close that ticket and post updates here. It is a user case (or habit) that I guess most people are used to. My own user case (desktop, no laptop, for medical reasons) is perhaps a bit of an outlier case.

**Comment by gaow** [comment]
Implemented and tested on small examples. I'll catch up with documentation tomorrow. To really use it properly one needs to have some familiarity of a set of SoS command utilities (that wrappers and extends eg PBS system specific utilities, and beyond). But documentation might be a challenge there, too.

**Comment by gaow** [comment]
But I do agree that the experience of sending jobs directly from the cluster is a lot smoother than sending from the local machine to the cluster, although no where to compare with just running on a local machine :P

**Comment by gaow** [comment]
@jean997 I've updated the documentation and the job template on this page:

https://stephenslab.github.io/dsc-wiki/FAQ/Remote_Computations.html

It is now default to submitting directly from the head node. Notice that there is minor syntax change to the template, so please copy this one and compare with your existing template to ensure yours is syntactically up to date.

To use this you'll have to update SoS, DSC and dscrutils all to the developer's version:

https://github.com/stephenslab/dsc#developer-notes

should be 3 lines of commands.

I've also polished an example here at the `dsc-reg` repo:

https://github.com/stephenslab/dsc-reg  (see `midway.yml`)

I ran this example with the `On-Host` method on midway. All looks okay. I've extracted from it some command examples and put that to the end of the wiki page too. 

Hope it helps. I'm leaving this ticket open for a while in case of related bug report.

**Comment by pcarbo** [comment]
@jean997 I'm still trying to understand your use case. Is this on the RCC cluster? Suppose you were to run `sbatch myscript.sh` in which `myscript.sh` contains a single command that runs `dsc mybenchmark.dsc -c 28`. This use all 28 CPUs on a single node to execute the DSC modules. Would this suffice?

**Comment by gaow** [comment]
@pcarbo Jean's case, in short, is to send `-c 28` to multiple nodes. And this `-c 28` is not 28 module instances like when we run on a local machine. It is 28 jobs each can have specified number of module instances, and go to different nodes. This is the cluster interface we've developed for DSC that were initially developed in SoS and ported over here. I've got a brief documentation on it and I updated it due to Jean's request.

Jean's trouble is that, initially I was under the impression I'm not supposed to run long time jobs on headnode, so in my initial version the submission part happens on a local machine. Jean finds it difficult to use for reasons she stated. And apparently my impression was not true (eg snakemake submitter can be a long running process). So now we have 2 versions as requested.
